packages/test/test-end-to-end-tests/src/test/incrementalSummaryContext.spec.ts

```diff
@@ -0,0 +1,243 @@
+import { IContainerRuntimeOptions } from "@fluidframework/container-runtime";
+import { IRequest } from "@fluidframework/core-interfaces";
+import {
+	IContainerRuntimeBase,
+	IIncrementalSummaryContext,
+	ISummaryTreeWithStats,
+	ITelemetryContext,
+} from "@fluidframework/runtime-definitions";
+import { requestFluidObject, SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import {
+	ITestFluidObject,
+	ITestObjectProvider,
+	TestFluidObjectFactory,
+	createSummarizerFromFactory,
+	summarizeNow,
+} from "@fluidframework/test-utils";
+import { describeNoCompat, getContainerRuntimeApi } from "@fluid-internal/test-version-utils";
+import { IFluidSerializer, SharedObject } from "@fluidframework/shared-object-base";
+import {
+	IChannelAttributes,
+	IChannelFactory,
+	IChannelServices,
+	IChannelStorageService,
+	IFluidDataStoreRuntime,
+} from "@fluidframework/datastore-definitions";
+import {
+	ISequencedDocumentMessage,
+	MessageType,
+	SummaryType,
+} from "@fluidframework/protocol-definitions";
+import { readAndParse } from "@fluidframework/driver-utils";
+import { pkgVersion } from "../packageVersion";
+
+// mark as experimental
+class TestSharedObjectFactory implements IChannelFactory {
+	public static readonly Type = "https://graph.microsoft.com/types/test-shared-object";
+
+	public static readonly Attributes: IChannelAttributes = {
+		type: TestSharedObjectFactory.Type,
+		snapshotFormatVersion: "0.1",
+		packageVersion: pkgVersion,
+	};
+
+	public get type(): string {
+		return TestSharedObjectFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return TestSharedObjectFactory.Attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<TestSharedObject> {
+		const sharedObject = new TestSharedObject(id, runtime, attributes, "TestSharedObject");
+		await sharedObject.load(services);
+		return sharedObject;
+	}
+
+	/**
+	 * {@inheritDoc @fluidframework/datastore-definitions#IChannelFactory.create}
+	 */
+	public create(document: IFluidDataStoreRuntime, id: string): TestSharedObject {
+		return new TestSharedObject(id, document, this.attributes, "TestSharedObject");
+	}
+}
+
+const snapshotFileName = "header";
+interface ISnapshot {
+	blobs: string[];
+}
+
+interface IBlob {
+	value: string;
+	seqNumber: number;
+}
+
+interface IOp {
+	type: "blobStorage";
+	value: string;
+}
+class TestSharedObject extends SharedObject {
+	static getFactory(): IChannelFactory {
+		return new TestSharedObjectFactory();
+	}
+	private readonly blobMap: Map<string, IBlob> = new Map();
+
+	protected summarizeCore(
+		serializer: IFluidSerializer,
+		telemetryContext?: ITelemetryContext | undefined,
+		incrementalSummaryContext?: IIncrementalSummaryContext | undefined,
+	): ISummaryTreeWithStats {
+		const builder = new SummaryTreeBuilder();
+
+		for (const [blobName, blobContent] of this.blobMap.entries()) {
+			if (
```

Comment:
This is the key piece of code that does incremental summary. Some comments here that explain how exactly that happens would be helpful.